src/iceberg/manifest_writer_internal.h

```diff
@@ -0,0 +1,160 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+/// \file iceberg/internal/manifest_writer_internal.h
+/// Writer implementation for manifest list files and manifest files.
+
+#include "iceberg/manifest_writer.h"
+#include "iceberg/v1_metadata.h"
+#include "iceberg/v2_metadata.h"
+#include "iceberg/v3_metadata.h"
+
+namespace iceberg {
+
+/// \brief Write manifest entries to a manifest file.
+class ManifestWriterImpl : public ManifestWriter {
+ public:
+  explicit ManifestWriterImpl(int64_t snapshot_id, std::unique_ptr<Writer> writer,
+                              std::shared_ptr<Schema> schema)
+      : schema_(std::move(schema)), writer_(std::move(writer)) {}
+
+ private:
+  std::shared_ptr<Schema> schema_;
+  std::unique_ptr<Writer> writer_;
+};
+
+/// \brief Write v1 manifest entries to a manifest file.
+class ManifestWriterV1 : public ManifestWriterImpl {
+ public:
+  explicit ManifestWriterV1(int64_t snapshot_id, std::unique_ptr<Writer> writer,
+                            std::shared_ptr<Schema> schema)
+      : ManifestWriterImpl(snapshot_id, std::move(writer), std::move(schema)) {}
+
+  Status WriteManifestEntry(const ManifestEntry& entry) const override;
+
+  Status Close() override;
+
+ private:
+  V1MetaData::ManifestEntryWrapper wrapper_;
+};
+
+/// \brief Write v2 manifest entries to a manifest file.
+class ManifestWriterV2 : public ManifestWriterImpl {
+ public:
+  explicit ManifestWriterV2(int64_t snapshot_id, std::unique_ptr<Writer> writer,
+                            std::shared_ptr<Schema> schema)
+      : ManifestWriterImpl(snapshot_id, std::move(writer), std::move(schema)),
+        wrapper_(snapshot_id) {}
+
+  Status WriteManifestEntry(const ManifestEntry& entry) const override;
+
+  Status Close() override;
+
+ private:
+  V2MetaData::ManifestEntryWrapper wrapper_;
+};
+
+/// \brief Write v3 manifest entries to a manifest file.
+class ManifestWriterV3 : public ManifestWriterImpl {
+ public:
+  explicit ManifestWriterV3(int64_t snapshot_id, int64_t first_row_id,
+                            std::unique_ptr<Writer> writer,
+                            std::shared_ptr<Schema> schema)
+      : ManifestWriterImpl(snapshot_id, std::move(writer), std::move(schema)),
+        wrapper_(snapshot_id) {}
+
+  Status WriteManifestEntry(const ManifestEntry& entry) const override;
+
+  Status Close() override;
+
+ private:
+  V3MetaData::ManifestEntryWrapper wrapper_;
+};
+
+/// \brief Write manifest files to a manifest list file.
+class ManifestListWriterImpl : public ManifestListWriter {
```

Comment:
The Java-equivalent has `protected abstract ManifestFile prepare(ManifestFile manifest)`, we need to add it too.